### ISSUE TITLE
Job wrapper tests & suggestions

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#236]
In GitLab by @ jlope2 on Feb 6, 2017, 17:36

This issue is to analyse how new job wrapper  feature works in autosubmit 3.8.x and to suggest new ideas for future developments.

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 6, 2017, 17:52

Hi I am adding some suggestions to improve the current vertical wrapper. It seems to work good for now, but because of unrelated issues with MN3 I had to resubmit all my jobs.

- add an option to enable/disable the wrapper for each job
- add a per-job max walltime (WRAPPER_WALLTIME), instead of (or in addition to)using MAX_WALLCLOCK in platforms_xxxx.conf
- add an option to specify the number of internal scripts, instead of having AS compute it
- allow to get (via email) and see (via autosubmit monitor) the "real" status of each job, if the wrapper script is running then the inner jobs could appear a completed, running or queuing
- show the wrapper script as a parent job of the inner jobs, or a box around the inner jobs, in the AS graph plot

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 7, 2017, 12:41

Hi @dmanubens @ jlope2 it is highly important that we can submit wrapper scripts with dependencies, which will help reduce waiting times. What is the status of this feature? I would need this this week! thanks

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 7, 2017, 13:38

another suggestion:
- add some output from the wrapper script like the time when each job starts and finishes, and perhaps the total time used. This will make it easier to fine-tune the walltime.

**Comment by kinow** [comment]
In GitLab by @ jlope2 on Feb 7, 2017, 15:10

Hi @etiennesky,

Regarding the status of the wrappers dependencies, it's almost ready.  I've solved the performance/usability problems within Autosubmit, now I'm waiting the answer of David Vicente about one question related with the LSF dependencies (it's working well but when something fails, then the scheduler does not kill those dependencies with conditions that cannot be satisfied). I hope solve these issues during the next hours.

Related with your last suggestion, some of this information is available in the STATs files. However we have a task in our TODO list to improve it. We will see how we adapt these STATs with the packages.

Thanks!

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 9, 2017, 12:48

Another suggestion, please disable the wrapper if there is only one job (e.g. the POST job after a SIM job). thanks!

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 9, 2017, 12:52

Oh and it would be nice to have the jobid of the wrapper script, as well as the number of the inner job in the filenames of the log files.

If not logs will be lost if the job is run again!

e.g. a0ez_19500101_fc0_4_SIM.cmd.3.err ->  a0ez_19500101_fc0_4_SIM.cmd.2697522.3.err

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 13, 2017, 11:36

I am writing here the steps required to use the current vertical wrapper with auto-ecearth

1) Install the development version of autosubmit

These steps are adapted from the wiki page https://earth.bsc.es/wiki/doku.php?id=library:python_venvs

```
ssh -X bscesautosubmit01
module purge
mkdir ~/venvs/as_dev
virtualenv ~/venvs/as_dev
source ~/venvs/as_dev/bin/activate
pip install https://earth.bsc.es/gitlab/es/autosubmit/repository/archive.zip?ref=develop
autosubmit -v
```

2) Edit you project files

as an example, see the files in /esnas/autosubmit/a0ez/conf

- platforms_xxxx.conf should contain in the  marenostrum3 section MAX_WALLCLOCK = 72:00 for prace accounts, or 48:00 for BSC32 accounts
- autosubmit_xxxx.conf : raise the values of these variables e.g. MAXWAITINGJOBS=100, TOTALJOBS = 100, SAFETYSLEEPTIME = 300
- jobs_xxxx.conf : make sure the walltime of your job is slightly less than the intended time (by 5 minutes), to make sure that the wrapper fits as many job as you intended

3) Run your experiment

```
ssh -X bscesautosubmit01
module purge
source ~/venvs/as_dev/bin/activate
autosubmit create expid
autosubmit recovery -s -all expid
autosubmit run expid
```

4) Check the wrapper script after it has been submitted. You will see the internal jobs listed in the script, make sure yo have as many as intended.

For example, my a0ez experiment has wrapper jobs of 72h with 5 internal jobs.

in /gpfs/scratch/pr1eaq00/pr1eaq14/a0ez/LOG_a0ez:

ASThread_1486143418_2032_5.cmd a0ez_19500101_fc0_1_SIM.cmd  a0ez_19500101_fc0_2_SIM.cmd  a0ez_19500101_fc0_3_SIM.cmd  a0ez_19500101_fc0_4_SIM.cmd   a0ez_19500101_fc0_5_SIM.cmd

and the wrapper script contains 5 internal jobs:

scripts = ['a0ez_19500101_fc0_1_SIM.cmd', 'a0ez_19500101_fc0_2_SIM.cmd', 'a0ez_19500101_fc0_3_SIM.cmd', 'a0ez_19500101_fc0_4_SIM.cmd', 'a0ez_19500101_fc0_5_SIM.cmd']

@obellprat @Eleftheriaexarchou @vsicardi

**Comment by kinow** [comment]
In GitLab by @Eleftheriaexarchou on Feb 13, 2017, 11:39

Thanks @etiennesky this is useful for @fmassonn too

**Comment by kinow** [comment]
In GitLab by @fmassonn on Feb 13, 2017, 11:48

Hi,

Thanks for the info. I'm not entirely clear how the job wrapper feature differs from the job array (es/autosubmit#223). Does one imply the other? Are the two going to co-exist eventually?

François

**Comment by kinow** [comment]
In GitLab by @ jlope2 on Feb 13, 2017, 12:06

Hi,

The main difference between wrappers and arrays is that a wrapper is represented as a single scheduler job (only once in the queue) while the arrays are represented as different scheduler jobs (one per element in the array, so one queue time each).

Moreover, the arrays approach is not compatible with the vertical packages while the wrapper is compatible with both (horizontal & vertical).

However,  currently we only have implemented the arrays approach and the vertical approach of the wrappers.

Thanks,

Joan

**Comment by kinow** [comment]
In GitLab by @fmassonn on Feb 13, 2017, 12:16

OK, thanks for this explanation!

François

**Comment by kinow** [comment]
In GitLab by @etiennesky on Feb 14, 2017, 10:30

For those who need to update the autosubmit to latest development version (at your own risk) the following commands are necessary

```
ssh -X bscesautosubmit01
module purge
source ~/venvs/as_dev/bin/activate
pip install --upgrade https://earth.bsc.es/gitlab/es/autosubmit/repository/archive.zip?ref=develop
autosubmit -v
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 8, 2019, 15:57

* [X]   add an option to enable/disable the wrapper for each job
* [X]   add a per-job max walltime (WRAPPER_WALLTIME), instead of (or in addition to)using MAX_WALLCLOCK in platforms_xxxx.conf
* [X]   add an option to specify the number of internal scripts, instead of having AS compute it
* [x]   allow to get (via email) and see (via autosubmit monitor) the "real" status of each job, if the wrapper script is running then the inner jobs could appear a completed, running or queuing
* [X]   show the wrapper script as a parent job of the inner jobs, or a box around the inner jobs, in the AS graph plot

**Comment by kinow** [comment]
In GitLab by @mcastril on Nov 18, 2019, 09:52

closed
